### DISCUSSION TITLE
Formatter "flickering" Question

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#1556] [Q&A]
I am using langium 3.1-next (the same was observed with langium 2.x and we had no issues originating from switching to 3.1-next so far. All other tests seems to pass :-) )

We had already have some issues when specifying formatter behavior which has conflicting requirements.
Therefore we build a test to reformat a model folder multiple times (2x) and check if we get a "stable" format in the end by comparing the result of the first run with the second run...

Now, we still have some "flickering" cases.
I observed that this behavior was influence by spaces in the original model file (no space after the `,`):
````
keyword1 name1 : reference1 (.reference2="text",.reference3=true)
````
This produced a flickering behavior: an extra newline is inserted after the first run
```
keyword1 name1 : reference1
(
     .reference2 = "msgId",

     .reference3 = true
)
```

**However** the following model resulted in a stable result (**a space was added after the comma**):
```
keyword1 name1 : reference1 (.reference2="text", .reference3=true)
```
Now we get a stable result as expected (an no extra newline in the first run):
```
keyword1 name1 : reference1
(
     .reference2 = "msgId",
     .reference3 = true
)
```

Is this something one could image to "disappear" the 4.x version? For us this would be ok to continue here later...

Note: probably it is not a "flickering", but a two-step process to settle down (not tested)

**Comment by cdietrich** [comment]
@goto40 afaik the fix is only in the final release

**Reply by goto40** [reply]
Thank you - I will retest once we have 3.1!

**Reply by cdietrich** [reply]
3.1 is already available

**Reply by cdietrich** [reply]
for us it fixed the flickering problem

**Reply by goto40** [reply]
Cool! I will try tomorrow - I think my IT-system needs some time to "allow" the external new version to come through :-( (I cannot install it at the moment - but I see that is is available on the internet)

**Comment by goto40** [comment]
I tried with 3.1 now: the problem persists, but I analyzed the effect further:
When I reformat 3 times, it remains stable between the second and third run. This means it requires a two-step formatting to "settle down" in my case...

(we are also still checking if it is an issue in our formatter)

@msujew : for me it "works" now - however I could try to mimic the effect in a minimal example if you wish (or do you think it is explainable with other known issues already?)

**Reply by msujew** [reply]
That definitely sounds like a different/new issue.

A repro example would be great, thank you!
